<!--
 Copyright (c) 2021  IBM Corporation

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
-->

<launch>
  <arg name="save_samples" default="false"/>
  <arg name="save_state" default="false"/>
  <arg name="convert_points" default="false"/>
  <arg name="convert_imu" default="false"/>
  <arg name="convert_esp32" default="false"/>
  <arg name="robot" default="rover"/>
  <arg name="wireless_topics" default="['/wireless/beacons','/wireless/wifi','/beacons','/esp32/wifi']"/>
  <arg name="rate" default="1.0"/>
  <arg name="start" default="0"/>
  <arg name="load_state_filename" default=""/>
  <arg name="delay" default="-1"/>
  <arg name="quit_when_rosbag_finish" default="false"/>

  <param name="/use_sim_time" value="true" />

  <arg name="scan" default="velodyne_scan" />
  <arg name="imu" default="imu/data"/>
  <arg name="points2" default="velodyne_points"/>
  <arg name="imu_temp" default="imu_temp/data"/>
  <arg name="points2_temp" default="velodyne_points_temp"/>

  <arg name="configuration_basename" default="cartographer_2d_mapping.lua"/> <!--  cartographer_2d_mapping.lua or cartographer_2d_mapping_localization.lua  -->
  <arg name="save_state_filename" default="" unless="$(arg save_state)"/>
  <arg name="save_state_filename" default="$(arg bag_filename).pbstream" if="$(arg save_state)"/>
  <arg name="start_trajectory_with_default_topics" default="$(eval load_state_filename=='')"/>

  <arg name="play_limited_topics" default="false"/>

  <arg name="arg_topics" value="--topics /velodyne_packets /$(arg points2) /$(arg scan) /$(arg imu) /beacons /wireless/beacons /wireless/wifi" if="$(arg play_limited_topics)"/>
  <arg name="arg_topics" value="" unless="$(arg play_limited_topics)"/>

  <!-- parameters to tf2_beacons_listener-->
  <arg name="save_empty_beacon_sample" default="true"/>

  <include file="$(find mf_localization_mapping)/launch/cartographer_2d_VLP16.launch">
    <arg name="robot" value="$(arg robot)"/>
    <arg name="scan" value="$(arg scan)"/>
    <arg name="points2" value="$(arg points2)"/>
    <arg name="imu" value="$(arg imu)"/>
    <arg name="configuration_basename" value="$(arg configuration_basename)"/>
    <arg name="load_state_filename" value="$(arg load_state_filename)"/>
    <arg name="save_state_filename" value="$(arg save_state_filename)"/>
    <arg name="start_trajectory_with_default_topics" value="$(arg start_trajectory_with_default_topics)"/>
  </include>

  <node name="rviz" pkg="rviz" type="rviz"
      args="-d $(find cartographer_ros)/configuration_files/demo_2d.rviz" />

  <!-- /velodyne_packets to /velodyne_points -->
  <include file="$(find mf_localization_mapping)/launch/includes/VLP16_points_cloud_nodelet_cartographer.launch" if="$(arg convert_points)">
    <arg name="scan" value="$(arg scan)"/>
  </include>

  <!-- rename imu frame_id -->
  <node pkg="mf_localization" type="imu_frame_renamer.py" name="imu_frame_renamer" output="screen" if="$(arg convert_imu)">
    <remap from="imu_in" to="$(arg imu_temp)"/>
    <remap from="imu_out" to="$(arg imu)"/>
  </node>

  <!-- convert esp32 wifi_scan_str -->
  <node pkg="wireless_scanner_ros" type="esp32_wifi_scan_converter.py" name="esp32_wifi_scan_converter" if="$(arg convert_esp32)">
    <!--remap from="esp32/wifi_scan_str" to="esp32/wifi_scan_str"/-->
    <remap from="wireless/wifi" to="esp32/wifi"/>
  </node>

  <!-- play -->
  <node name="playbag" pkg="rosbag" type="play" required="$(arg quit_when_rosbag_finish)" if="$(eval delay == -1)"
	args="--pause -r $(arg rate) --start $(arg start) --clock $(arg bag_filename) $(arg arg_topics)" output="screen">
    <remap from="$(arg imu)" to="$(arg imu_temp)" if="$(arg convert_imu)"/>
    <remap from="$(arg points2)" to="$(arg points2_temp)" if="$(arg convert_points)"/>
  </node>

  <node name="playbag" pkg="rosbag" type="play" required="$(arg quit_when_rosbag_finish)" if="$(eval delay != -1)"
	args="-d $(arg delay) -r $(arg rate) --start $(arg start) --clock $(arg bag_filename) $(arg arg_topics)" output="screen">
    <remap from="$(arg imu)" to="$(arg imu_temp)" if="$(arg convert_imu)"/>
    <remap from="$(arg points2)" to="$(arg points2_temp)" if="$(arg convert_points)"/>
  </node>

  <!-- write samples json -->
  <node name="tf2_beacons_listener" pkg="mf_localization"
        type="tf2_beacons_listener.py" if="$(arg save_samples)">
    <param name="output" value="$(arg bag_filename).loc.samples.json" />
    <param name="topics" value="$(arg wireless_topics)"/>
    <param name="save_empty_beacon_sample" value="$(arg save_empty_beacon_sample)"/>
  </node>

</launch>
